File: package/main/src/Predicate/not.ts
/**
 * Creates a predicate that negates the given predicate
 * @param {(...arguments_: T) => boolean} function_ - The predicate to negate
 * @returns {(...arguments_: T) => boolean} A new predicate that returns the opposite
 * @example
 * const isEven = (n: number) => n % 2 === 0;
 * const isOdd = not(isEven);
 * isOdd(3); // true
 */
export const not =
  <T extends unknown[]>(
    function_: (...arguments_: T) => boolean,
  ): ((...arguments_: T) => boolean) =>
  (...arguments_: T): boolean =>
    !function_(...arguments_);

File: package/umt_python/src/predicate/not_.py
from collections.abc import Callable
from typing import ParamSpec

P = ParamSpec("P")


def not_(predicate: Callable[P, bool]) -> Callable[P, bool]:
    """
    Creates a predicate that negates the given predicate

    Args:
        predicate: The predicate to negate

    Returns:
        A new predicate that returns the opposite

    Example:
        >>> is_even = lambda n: n % 2 == 0
        >>> is_odd = not_(is_even)
        >>> is_odd(3)
        True
    """

    def negated(*args: P.args, **kwargs: P.kwargs) -> bool:
        return not predicate(*args, **kwargs)

    return negated
